#+begin_src latex
  \section*{Introduction}
  In this Tutorial we construct an \ty{sqlite} database from the NCBI
  taxonomy data. Then we query the database to find the taxon-IDs for
  our target organism. Based on these taxon-IDs, we carry out the search
  for target and neighbor genomes using \ty{neighbors}. We subject these
  genomes to two connected analyses, phylogeny reconstruction and
  identification of diagnostic markers.
  \section*{Construct Database}
  We download the taxonomy database and the genome reports from the
  NCBI.
#+end_src
#+begin_src sh <<construct.sh>>=
  wget ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz
  wget ftp.ncbi.nlm.nih.gov/genomes/GENOME_REPORTS/prokaryotes.txt
  wget ftp.ncbi.nlm.nih.gov/genomes/GENOME_REPORTS/eukaryotes.txt
  wget ftp.ncbi.nlm.nih.gov/genomes/GENOME_REPORTS/viruses.txt
#+end_src
#+begin_src latex
  We unpack the taxonomy database.
#+end_src
#+begin_src sh <<construct.sh>>=
  tar -xvzf taxdump.tar.gz
#+end_src
#+begin_src latex
  This generates nine files, of which we use two, \ty{nodes.dmp} with
  the nodes of the taxonomic tree, and \ty{names.dmp} with the taxon
  names. We construct the taxonomy database, \ty{tax.db}, by running
  \ty{makeNeiDb} in default mode. This takes less than a minute.
#+end_src
#+begin_src sh <<construct.sh>>=
  makeNeiDb
#+end_src
#+begin_src latex
  \section*{Query Database}
  The next step is to find the taxon-IDs for our target organism or
  target organisms. Say, our target organism is \emph{Aerococcus
  urinae}. We look up its taxon-ID using \ty{taxi}.
#+end_src
#+begin_src sh <<query.sh>>=
  taxi "Aerococcus urinae" tax.db
#+end_src
#+begin_src latex
  \begin{verbatim}
  # ID    Parent  Name
    1376  1375    Aerococcus urinae
  \end{verbatim}
  The result tells us that \emph{Aerococcus urinae} has taxon-ID 1376
  and its parent has taxon-ID 1375. To get an overview of the
  \emph{Aerococcus} clade, we draw it with our program for drawing
  trees, \ty{dree}.
#+end_src
#+begin_src sh <<query.sh>>=
  dree 1375 tax.db
#+end_src
#+begin_src latex
  Unfortunately, this tree is so large, it's illegible. So we restrict
  it to the clades with sequenced genomes.
#+end_src
#+begin_src sh <<query.sh>>=
  dree -g 1375 tax.db | dot -T x11
#+end_src
#+begin_src latex
   In Figure~\ref{fig:au1} the
  \emph{Aerococcus} taxa with sequenced genomes are colored.
  \begin{figure}
    \begin{center}
      \resizebox{\textwidth}{!}{\includegraphics{../tutorial/1375_t_g.ps}}
    \end{center}
    \caption{The taxonomy of \emph{Aerococcus} (1375) restricted to taxa
      with sequenced genomes, which are shown as colored
      nodes.}\label{fig:au1}
  \end{figure}
  However, taxon-IDs are difficult to interpret, so we convert them to
  taxon names in Figure~\ref{fig:au2}.
#+end_src
#+begin_src sh <<query.sh>>=
  dree -n -g 1375 tax.db | dot -T x11
#+end_src
#+begin_src latex
  \begin{figure}
    \begin{center}
      \resizebox{\textwidth}{!}{\includegraphics{../tutorial/1375_n_g.ps}}
    \end{center}
    \caption{The taxonomy tree of \emph{Aerococcus} with sequenced
      genomes and taxon names.}\label{fig:au2}
  \end{figure}
  Now we search for the neighbors of \emph{Aerococcus}.
#+end_src
#+begin_src sh <<query.sh>>=
  printf "1376\n" | neighbors tax.db
#+end_src
#+begin_src latex
  \begin{verbatim}
  # MRCA(targets): 1376, Aerococcus urinae
  # MRCA(targets+neighbors): 1375, Aerococcus
  # Type  Taxon-ID  Name                      Genomes
  t       1376      Aerococcus urinae         CP014161|CP065662
  tt      866775    Aerococcus urinae ACS     CP002512
  tt      1216979   Aerococcus urinae NBRC    -
  n       1377      Aerococcus viridans       CP014164
  n       51665     Aerococcus urinqeequi     CP013988|CP014162|...
  n       79233     Aerococcus sp. LV65.5:W1  -
  n       87541     Aerococcus christensenii  CP014159
  n       119206    Aerococcus sanguinicola   CP014160
  n       128944    Aerococcus urinaehominis  CP014163
  ...
  \end{verbatim}
  The output of \ty{neighbors} begins with three hashed lines. The first
  states the most recent common ancestor of the targets,
  \emph{A. urinae} with taxon-ID 1376. The second states the most recent
  common ancestor of the neighbors and the targets, \emph{Aerococcus},
  with taxon-ID 1375. The third hashed row is the header of the
  subsequent table, which consists of four columns, type, taxon-ID,
  name, and genomes. There are three possible types, ``t'' for known
  target, ``tt'' for new target, and ``n'' for neighbor. We see that
  there are two new targets, 866775 with one genome sequence, which we
  already saw in Figure~\ref{fig:au} and 1216979 without a genome
  sequence. We ignore plasmid sequences, which leaves seven genomes
  among the neighbors. We fetch them using a loop over their accessions,
  mark their names with prefix ``n'', and store them in the directory
  \ty{neighbors}.
#+end_src
#+begin_src sh <<query.sh>>=
  mkdir neighbors
  for a in CP014164 CP013988 CP014162 CP063065
	   CP014159 CP014160 CP014163
  do
      efetch -db nucleotide -format fasta \
	     -id $a > neighbors/n${a}.fasta
  done
#+end_src
#+begin_src latex
  We compute a quick phylogeny for the target and neighbor genomes using
  \ty{phylonium}\footnote{\ty{github.com/evolbioinf/phylonium}} and
  programs form the
  \ty{biobox}\footnote{\ty{github.com/evolbioinf/biobox}}. The result in
  Figure~\ref{fig:auTree} shows that the targets are monophyletic, which
  is necessary for a successful \ty{fur} run.
  \begin{figure}
    \begin{center}
      \scalebox{0.75}{\includegraphics{../tutorial/au}}
    \end{center}
    \caption{Phylogeny of target and neighbor genomes; targets are
      prefixed ``t'', neighbors ``n''.}\label{fig:auTree}
  \end{figure}
#+end_src
#+begin_src sh <<query.sh>>=
  phylonium targets/* neighbors/* | nj | midRoot | plotTree
#+end_src
#+begin_src latex
  Now we can construct the \ty{fur} database with \ty{makeFurDb} and run
  \ty{fur}\footnote{\ty{github.com/evolbioinf/fur}} on it to get
  approximately 474 kb potential marker material. The program \ty{cres}
  used to quantify the output of \ty{fur} is also part of the
  \ty{biobox}.
#+end_src
#+begin_src sh <<query.sh>>=
  makeFurDb -t targets/ -n neighbors/ -d au.db
  fur -d au.db | cleanSeq > au.fasta
  cres au.fasta
#+end_src
#+begin_src latex
  If we look again at the phylogeny in Figure~\ref{fig:auTree}, we might
  want to rethink the position of CP002512. Perhaps it should be a
  neighbor rather than a target? This of course depends on the details
  of our test system. But let's say we make CP002512 a neighbor; then
  our set of potential markers shrinks to 105 kb.
#+end_src
#+begin_src sh <<query.sh>>=
  mv targets/tCP002512.fasta neighbors/
  makeFurDb -t targets/ -n neighbors/ -d au2.db
  fur -d au2.db/ | cleanSeq > au2.fasta
  cres au2.fasta
#+end_src
#+begin_src latex
  We might extend this analysis by making CP002512 the target and
  everything else the neighors. This returns 190 kb marker material.
#+end_src
#+begin_src sh <<query.sh>>=
  mv targets/* neighbors/
  mv neighbors/tCP002512.fasta targets/
  makeFurDb -t targets/ -n neighbors/ -d au3.db
  fur -d au3.db/ | cleanSeq > au3.fasta
  cres au3.fasta
#+end_src
#+begin_src latex
  \subsection*{Pick Clades from Phylogeny}
  Sometimes a tree of genomes is very large, making it inconvenient to
  manually list the taxa that belong to a particular clade. This isn't
  an issue in our analysis (Figure~\ref{fig:auTree}), but if it were, we
  could use the programs \ty{lads} and \ty{pickle} to pick individual
  clades. \ty{lads} labels the nodes in a tree, yielding
  Figure~\ref{fig:auTree2}.
  \begin{figure}
    \begin{center}
      \scalebox{0.75}{\includegraphics{../tutorial/au2}}
    \end{center}
    \caption{Phylogeny of target and neighbor genomes with internal
      nodes, or \emph{clades}, labeled.}\label{fig:auTree2}
  \end{figure}
#+end_src
#+begin_src sh <<query.sh>>=
  lads au.nwk > au2.nwk
  plotTree au2.nwk
#+end_src
#+begin_src latex
  Now we can pick a clade, say clade 6, and print its leaf labels with
  \ty{pickle}.
#+end_src
#+begin_src sh <<query.sh>>=
  pickle au2.nwk
#+end_src
#+begin_src latex
  \begin{verbatim}
  # Selected clades
  ## 6
  nCP063065
  nCP014164
  nCP013988
  nCP014162
  \end{verbatim}
#+end_src
