Diagnostic PCR-markers are designed to amplify all members of a target
set of organisms and nothing else. A powerful first step to ensuring
marker specificity is to compare the target genomes to the genomes of
the closest distinct relatives, the neighbors. This usually removes
the vast majority of non-specific material. The small remainder can
then be further tested by \emph{in silico} PCR against, say, the
non-redundant collection of nucleotide sequences, \ty{nt}. The
program \ty{fur} implements this comparison between targets and
neighbors for finding marker candidates. Markers constructed from its
output have excellent specificity and sensifivity~\cite{hau21:fur}.

Users of programs like \ty{fur} need to know the neighbors of their
targets. But how can neighbors be discovered, if they aren't already
known? To answer this question, consider the toy taxonomy in
Figure~\ref{fig:tax}, where the numbers are taxon-IDs that are linked
to genome accessions. Let taxa 7 and 4 be our targets shown in
bold. Their most recent common ancestor is 3. This implies there are
three additional targets, 3, 5, and 6. The neighbors are the nodes in
the subtree rooted on 3's parent, minus the parent and minus the
targets. So in our example there are five neighbors, 2, 8, 9, 10, and
11. Notice that we look up all nodes in a subtree, not just the
leaves, as genome sequences might be associated with taxa both in
terminal and internal nodes.

\begin{figure}
\begin{center}
\input{tree}
\end{center}
\caption{Toy taxonomy for finding targets and neighbors; the original
targets are shown in bold.}\label{fig:tax}
\end{figure}

To put this a bit more formally, let $m$ be the most recent common
ancestor of the targets in the taxonomy. Their neighbors,
$\mathcal{N}$, are then computed by subtracting the nodes in $m$'s
subtree from the nodes in its parent's subtree, minus the parent
\begin{equation}\label{eq:nei}
\mathcal{N} = s(p(m)) - s(m) - p(m),
\end{equation}
where $s(v)$ returns the nodes in the subtree rooted on $v$, and
$p(v)$ the parent of $v$. This set subtraction is implemented in the
program \ty{neighbors}.

The Neighbors package consists of the ten programs listed in
Table~\ref{tab:pro}. Five of these programs are based on the taxonomy,
four on a phylogeny, and one, \ty{outliers}, on numerical data.

\begin{table}
\caption{The ten programs of Neighbors}\label{tab:pro}
\begin{center}
\begin{tabular}{rlll}
\hline
\# & Name & Based on & Function\\\hline
1 & \ty{ants} & taxonomy & list ancestors\\
2 & \ty{climt} & phylogeny & climb tree\\
3 & \ty{dree} & taxonomy & draw tree\\
4 & \ty{fintac} & phylogeny & find target clade\\
5 & \ty{land} & phylogeny & label nodes\\
6 & \ty{makeNeiDb} & taxonomy & make neighbors database\\
7 & \ty{neighbors} & taxonomy & find neighbors (and targets)\\
8 & \ty{outliers} & numbers & find outliers\\
9 & \ty{pickle} & phylogeny & pick clades\\
10 & \ty{taxi} & taxonomy & get taxon-ID for taxon name\\\hline
\end{tabular}
\end{center}
\end{table}

The taxonomy is supplied as an \ty{sqlite} database, let's call
it \ty{neidb}, which is built using the program \ty{makeNeiDb}. As
shown in Figure~\ref{fig:db}, \ty{neidb} consists of three
tables, \ty{genome}, \ty{taxon}, and \ty{genome\_count}.

\begin{figure}[ht]
  \begin{center}
  \LARGE
  \input{labels}
    \scalebox{0.55}{\includegraphics{db}}
  \end{center}
  \caption{Diagram of \ty{neidb}.}\label{fig:db}
\end{figure}

Each \ty{genome} has a unique accession, which also serves as primary
key, comes from an organism identified by its taxon-ID, has a size,
and an assembly level. Each genome belongs to a \ty{taxon}. A taxon
has a unique taxon-ID, which serves as primary key. A taxon also has a
parent, a rank, a name, a common name, and a score.

The table \ty{genome\_count} stores genome counts for each taxon
identified by its taxid. Genome counts are broken down by assembly
level, which can be complete, chromosome, scaffold, or contig, in that
order of quality. Genome counts are further subdivided into raw and
recursive. A raw count is just the number of genomes listed in
table \ty{genome} for that taxon. A recursive count it the number of
genomes in the subtree rooted on that taxon. This recursive count
cannot be conveniently implemented in SQL, so we implement it in Go.

The table \ty{image} stores images of taxa. Each image has an ID, a
URL, and an attribution. Since a taxon might have one or more images,
and an image might depict more than one taxon, tables \ty{image}
and \ty{taxon} have a many to many relationship.

An important function of \ty{neidb} is to support text searches of
taxon names. These searches should allow client programmers to write
interfaces modeled on Google, where the user starts typing and
instantaneously gets a proposal list of matches. After typing the
first character, the number of possible matches will usually be much
larger than can sensibly be passed to the user. So we need to order
the matches by relevance. Google's initial breakthrough was to rank
matching web sites by the number of incoming links. We propose to rank
taxa by the number of genome sequences, which we store as attribute
score in table \ty{taxon}.

Once the database is constructed, we can query it. The tutorial shows
how to do that for \emph{Leginonella pneumophila}. It is a notorious
water-borne pathogen that can cause pneumonia in humans. The
program \ty{taxi} gives us the taxon-ID for our focal subspecies
of \emph{L. pneumophila}, \emph{subsp. pneumophila}.

It is often useful to place this taxon-ID into context, for which we
have two programs, \ty{dree} and \ty{andi}. Starting from a
taxon-ID, \ty{dree} draws the taxonomic tree rooted on that
taxon-ID. Instead of walking from a given node towards the leaves, the
program \ty{ants} starts from a taxon and walks in the opposite
direction toward the universal root. Along this path, \ty{ants} lists
all ancestral taxa of our focal taxon. Once we've got our bearings in
the taxonomy, we can query it with \ty{neighbors} to get the complete
set of target and neighbor genomes currently available.

The genomes returned by \ty{neighbors} form the raw material for
marker discovery. However, the classification into targets and
neighbors retrieved from the taxonomy may contradict the phylogeny
calculated from the target and neighbor genomes to determine the final
list of target and neighbor genomes. So we compute a phylogeny from
our target and neighbor genomes. A program for doing this efficiently
is \ty{phylonium}~\cite{klo20:phy}.

The phylogenies of targets and neighbors may comprise hundreds of
taxa. To help analyze such large phylogenies, Neighbors
contains \ty{land} for labeling nodes, \ty{pickle} for picking nodes,
\ty{fintac} for finding the target clade, and \ty{climt} for climbing
the phylogenetic tree.

Even a clean set of phylogenetic targets might still contain genomes
that are outliers in some way, for example with respect to their
genome lengths. The program \ty{outliers} helps find such
outliers.
